### PR TITLE
Slate integration

### PR DESCRIPTION
**Pull request by tiberiuichim**

### CODE REVIEW

**Comment by tiberiuichim** [comment]
Listing here things that I've ripped out and will have to be added back:

- `editor/plugins/Link` 
- HashLink in `Link/AppExtras/HashLink`
- BlocksBrowserWidget
-

**Comment by sneridagh** [comment]
Can we name it @plone/volto-slate? 😄 Is it much trouble? It comes to my mind a couple of situations that would come in handy and avoid confusion.

File: src/config/index.js [hunk 1]
```diff
@@ -182,6 +186,7 @@ let config = {
   components,
 };
 
+config = simpleLink(minimalDefault(applySlateConfiguration(config)));
```
Comment (sneridagh) [inline src/config/index.js:189]: Why do we need this explicitly and not allow slate to set it by itself? As an add-on?

**Review by sneridagh — COMMENTED** [review]
Comments and questions for now, and wish list. We would like to have a list with all of them. I will try to arrange it, so we can have them all centralised.

BTW, good work. From my point of view, it only needs polish and wax, and adapt it to make believe volto-slate that it's now in core and it's the default :) 

We will need a look at docs, replace draft everywhere. Deprecation notice in the Upgrade Guide, and upgrade guide itself. Also document the plugins, and what they do, how they enable them...

/cc @pbauer anything that we already have that we could include already in the docs? Specially the story Volto draftJS -> Volto Slate

Let's try to keep momentum! Gogogo

File: cypress/support/volto-slate.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
+export const slateBeforeEach = (contentType = 'Document') => {
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply cypress/support/volto-slate.js:1]: This is used to autologin and create a page before each volto-slate test.

File: .eslintrc [hunk 3]
```diff
@@ -32,14 +32,19 @@
   "settings": {
     "import/resolver": {
       "alias": {
-        "map": [["@plone/volto", "./src"], ["@package", "./src"], ["@root", "./src"]],
+        "map": [
+          ["@plone/volto", "./src"],
+          ["@package", "./src"],
+          ["@root", "./src"],
+          ["volto-slate", "./packages/volto-slate/src"]
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply .eslintrc]: Alias updated.

File: .github/workflows/acceptance.yml [hunk 4]
```diff
@@ -131,6 +131,70 @@ jobs:
           name: cypress-videos
           path: cypress/videos
 
+  corevoltoslate:
+    runs-on: ubuntu-latest
+    name: Core Volto Slate
+    strategy:
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply .github/workflows/acceptance.yml:138]: Timeout added.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 ## 15.8.1 (unreleased)
 
+- Ship Volto with volto-slate included. @tiberiuichim @razvanmiu @eea
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply CHANGELOG.md]: I've changed the changelog line.

File: cypress/tests/core/volto-slate/02-block-slate.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { slateBeforeEach, slateAfterEach } from '../../../support';
+
+describe('Block Tests', () => {
+  beforeEach(slateBeforeEach);
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply cypress/tests/core/volto-slate/02-block-slate.js:4]: I've removed the 'slateAfterEach' function but I kept the 'slateBeforeEach'.

File: packages/volto-slate/cypress.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/cypress.json]: Cypress removed from packages/volto-slate.

File: packages/volto-slate/package.json [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  "keywords": [
+    "volto-addon",
+    "volto-slate",
+    "volto-slate-addon",
+    "volto",
+    "plone",
+    "react"
+  ],
+  "addons": [
+    "@eeacms/volto-object-widget"
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/package.json]: I've added the widget used from '@eeacms/volto-object-widget' into packages/volto-slate. Maybe I should move it somewhere else?

File: packages/volto-slate/package.json [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    "@cypress/code-coverage": "^3.9.5",
+    "@testing-library/react": "9.5.0",
+    "@types/jest": "^25.2.3",
+    "babel-plugin-transform-class-properties": "^6.24.1"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git@github.com:eea/volto-slate.git"
+  },
+  "scripts": {
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/package.json:31]: Done.

File: packages/volto-slate/src/blocks/Description/index.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import descriptionSVG from '@plone/volto/icons/description.svg';
+import TitleBlockView from '../Title/TitleBlockView';
+import TitleBlockEdit from '../Title/TitleBlockEdit';
+
+export default (config) => {
+  const className = 'documentDescription';
+  const formFieldName = 'description';
+
+  config.blocks.blocksConfig.description = {
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/src/blocks/Description/index.js]: I've added the title and description block in core. I still have to add Table block and Table of Content (ToC) block.

File: packages/volto-slate/src/components/ElementEditor/ContextButtons.jsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/src/components/ElementEditor/ContextButtons.jsx]: This has some helpers functions used to create complex plugins (like the Link plugin) which have their own configuration.

File: packages/volto-slate/src/components/themes/View/DefaultSlateView.jsx [hunk 12]
```diff
@@ -0,0 +1,123 @@
+/**
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/src/components/themes/View/DefaultSlateView.jsx]: Removed.

File: packages/volto-slate/src/editor/plugins/index.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import installBlockQuotePlugin from './Blockquote';
+import installImage from './Image';
+// import installLinkPlugin from './Link';
```
[the remark this answers is not in the document]
Comment (razvanMiu) [reply packages/volto-slate/src/editor/plugins/index.js]: I've used the "Link" plugin and removed the "SimpleLink" one.

**Comment by giuliaghisini** [comment]
instead of including and repeat code in each Edit component for blocks.. is it possible to introduce a widget to simplify imports and avoiding some eventually distributed bugs?

**Comment by tiberiuichim** [comment]
@giuliaghisini Is this what you're thinking? https://github.com/plone/volto/blob/slate-integration/packages/volto-slate/src/widgets/RichTextWidget.jsx

**Comment by giuliaghisini** [comment]
> @giuliaghisini Is this what you're thinking? https://github.com/plone/volto/blob/slate-integration/packages/volto-slate/src/widgets/RichTextWidget.jsx

yes! so.. why we don't use it in existing blocks instead including slate directly?

**Comment by tiberiuichim** [comment]
@giuliaghisini Title/Description blocks have a different behavior (need to render as a specific tag, only one line, etc). The Table also has specific requirements (navigation keys, etc).

**Review by nileshgulia1 — COMMENTED** [review]
@sneridagh @tiberiuichim I took care of the cypress tests. I think we have to publish `@plone/volto-slate` to npm. As it can't able to resolve paths in project-generator.
 
`Module not found: Can't resolve '@plone/volto-slate' in '/home/runner/work/volto/volto/my-volto-app/node_modules/@plone/volto/src/config'`

**Comment by tiberiuichim** [comment]
@nileshgulia1 No, `@plone/volto-slate` should be provided as an alias by Volto. Maybe Volto needs to include the packages folder, when installed? Something is not working properly, Volto includes `@plone/volto-slate`.

**Comment by nileshgulia1** [comment]
@tiberiuichim I don't have a strong opinion if we should ship `packages/` folder with volto. We have the [`@plone/scripts`](https://github.com/plone/volto/blob/master/package.json#L234) package which is installed as a dependency in Volto, but not the case here with `@plone/volto-slate`. I think we should make a decision on it soon.

**Comment by tiberiuichim** [comment]
@nileshgulia1 That decision was made at the Bethoven sprint, together with @sneridagh and @robgietema. The existing layout of the packages in this PR will be kept, `@plone/volto-slate` will be included with Volto, but it won't be published as a separate package.

**Review by sneridagh — APPROVED** [review]
LGTM, only a couple of heads up.

**Comment by sneridagh** [comment]
@tiberiuichim @razvanMiu @nileshgulia1 I assume this is ready? I realised that there are still two commented Cypress tests.

@plone/volto-team if anyone else could take a look, it would be great! More eyes are always welcome.

We are almost there!

**Comment by sneridagh** [comment]
mmm it seems the generator tests have problems.

**Comment by sneridagh** [comment]
Now, it should be green :)

**Comment by sneridagh** [comment]
💚

**Review by nileshgulia1 — APPROVED** [review]
LGTM

File: packages/volto-slate/locales/de/LC_MESSAGES/volto.po [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+#: editor/plugins/SimpleLink/index
+msgid "Add link"
+msgstr "Link hinzufügen"
+
+#: blocks/Table/TableBlockEdit
+msgid "Align text"
+msgstr ""
+
+#: editor/plugins/StyleMenu/StyleMenu
```
Comment (ksuess) [inline packages/volto-slate/locales/de/LC_MESSAGES/volto.po:26]: Left over from former StyleMenu. more below.

**Review by ksuess — APPROVED** [review]
A project that already is on Slate with customizations (schema enhanced block, toolbar, LinkElement with `/@@download` and anchor) and addon '@rohberg/volto-slate-glossary` is up and running with new `@plone/volto-slate`, yeah! 
The only thing that I am missing, is the `config.settings.slate.styleMenu.inlineStyles`, but this can be brought back in an add-on, should be easy.

File: packages/volto-slate/locales/de/LC_MESSAGES/volto.po (hunk 14 above)
Comment (sneridagh) [reply packages/volto-slate/locales/de/LC_MESSAGES/volto.po:26]: @tiberiuichim @razvanMiu did we remove the StyleMenu? It seems there are this leftovers

Comment (sneridagh) [reply packages/volto-slate/locales/de/LC_MESSAGES/volto.po]: Updated.

**Comment by nileshgulia1** [comment]
@sneridagh @ksuess The commit https://github.com/plone/volto/pull/3313/commits/3b149d91fab6a7c2498c0b5d2ab7cd437d238189
alters volto-slate to add a slate text block from blocksChooser. 
Right now if you add a new block from BlockChooser it adds as a draftjs one.

**Comment by sneridagh** [comment]
yeah @nileshgulia1 @ksuess I wonder how cypress tests are working then? It should be ok.

could you please take care and test it? Thanks!

File: src/config/index.js [hunk 15]
```diff
@@ -185,6 +186,7 @@ let config = {
   components,
 };
 
+config = applySlateConfiguration(config);
```
Comment (sneridagh) [inline src/config/index.js:189]: ups, it's a typo, right?

Comment (sneridagh) [reply src/config/index.js:189]: @tiberiuichim the default is slate, but the defaultBlockType remains in "text" in the config. How could that be? What is the thing that forces it? Also, it's important that we honour this setting, in the use case that someone wants to remain in "draftJS" world in Plone 6 until a migration is figured out.

**Comment by sneridagh** [comment]
@tiberiuichim @razvanMiu Uh, sure, I'm idiot, if you need this, then you do that in your add-on/project, then it will be overrided.

**Comment by razvanMiu** [comment]
> @sneridagh @ksuess The commit [3b149d9](https://github.com/plone/volto/commit/3b149d91fab6a7c2498c0b5d2ab7cd437d238189) alters volto-slate to add a slate text block from blocksChooser. Right now if you add a new text block from BlockChooser it adds as a draftjs one.

@sneridagh I've made this change in [here](https://github.com/plone/volto/commit/0fbaed97ea80196954d0e4bee2b07ec71af68558#diff-97feebaf386a564c904d56c942e34b479f1a0b4b6175feb765d4e75e5f98a040R118) only for testing something and I forgot to revert it.
If someone wants to use draft js then in an addon this settings needs to be added:
```
config.settings.defaultBlockType = 'text'
config.blocks.blocksConfig.text.restricted = false
config.blocks.blocksConfig.table.restricted = false
config.blocks.blocksConfig.slate.restricted = true
config.blocks.blocksConfig.slateTable = true
```
Only the title & description blocks would still use the slate block.
The ToC is compatible with draftjs.

**Review by razvanMiu — APPROVED** [review]
LGTM

**Comment by tiberiuichim** [comment]
@ksuess @sneridagh I'm in favour of including the Style menu. The only change I would do, is to use a button as a trigger, so it would take less space in the toolbar

**Comment by sneridagh** [comment]
Please excuse my ignorance, the StyleMenu is the one on the toolbar that is a dropdown with the name of the available styles? I'm ok to add it in the codebase, but I'd leave it as opt-in. What do you think?

**Comment by sneridagh** [comment]
I think that people would modify the toolbar anyways per project, depending in they need some of the buttons or not.

**Comment by ksuess** [comment]
Yes, sure the StyleMenu should be optional.
I could imagine: if a project does https://github.com/eea/volto-slate/blob/2e77a2bec56201a5ff9791e6aad9e8fa95e8d3ab/src/editor/plugins/StyleMenu/index.js#L15-L26 then a button appears to open a dropdown. With @tiberiuichim, this would take as small space as possible and would look still slim. Will do a proposal in https://github.com/plone/volto/issues/3494
if you all aggree to bring it back. If not, I am completely happy to do it in an add-on.

**Comment by tiberiuichim** [comment]
@ksuess @sneridagh My vote is for the Style Menu to come with Volto, disabled.

The style menu covers 2 potential holes for a project:

- I have a big number of custom styles
- I don't have the time/resources to design custom icons for each additional button I want to use.

**Comment by ksuess** [comment]
> How do we proceed? Do we merge now, then address it in another PR?

I created a new issue for bringing back the StyleMenu. This PR should not be blocked by this side topic.

**Pull request merged**